File: cru_cl2_examples.R
library(ncdf4)
library(terra)
library(rnaturalearth)
library(rnaturalearthdata)

# Download cru_cl2.nc from Zenodo (DOI to be added after upload)
# and place it in your working directory, then run this script.

ncfile <- "cru_cl2.nc"
nc <- nc_open(ncfile)
print(nc)  # inspect all variables and dimensions

lons <- ncvar_get(nc, "lon")
lats <- ncvar_get(nc, "lat")
world <- ne_countries(scale = "medium", returnclass = "sf")

# --- Example 1: Global map of mean July temperature ---

tmp_rast <- rast(ncfile, subds = "tmp")   # terra reads coordinates directly
plot(tmp_rast[[7]], main = "CRU CL 2.0 — Mean temperature July (°C)")
plot(world$geometry, add = TRUE, border = "white", lwd = 0.5)


# --- Example 2: Point extraction — all variables at one location ---

target_lon <- 133.8807  # Alice Springs, Australia
target_lat  <- -23.6980
xy <- cbind(target_lon, target_lat)

monthly_vars <- c("pre", "rd0", "frs", "tmp", "dtr", "reh", "sunp", "wnd")

point <- data.frame(month = 1:12)
for (v in monthly_vars) {
  r_var      <- rast(ncfile, subds = v)
  point[[v]] <- as.numeric(terra::extract(r_var, xy)[1, ])
}
elv_r     <- rast(ncfile, subds = "elv")
point$elv <- as.numeric(terra::extract(elv_r, xy)[1, 1])

# Derive min/max temperature from tmp and dtr
point$tmin <- point$tmp - point$dtr / 2
point$tmax <- point$tmp + point$dtr / 2

print(point)

par(mfrow = c(2, 1))
plot(point$month, point$tmax, type = "l", col = "red",
     ylim = range(point$tmin, point$tmax),
     xlab = "Month", ylab = "Temperature (°C)",
     main = paste("Alice Springs temperature range (elev", round(point$elv[1]), "m)"))
lines(point$month, point$tmin, col = "blue")
legend("topright", c("Tmax", "Tmin"), col = c("red", "blue"), lty = 1)

plot(point$month, point$pre, type = "h", col = "steelblue",
     xlab = "Month", ylab = "Precipitation (mm)",
     main = "Alice Springs precipitation")


# --- Example 3: All months as a terra SpatRaster ---

nc_close(nc)

pre_rast <- rast(ncfile, subds = "pre")
names(pre_rast) <- month.abb
plot(pre_rast[[c(1, 4, 7, 10)]],
     main = paste("Precipitation", c("Jan", "Apr", "Jul", "Oct"), "(mm)"))

File: cru_cl2_examples.jl
using NCDatasets

ncfile = "cru_cl2.nc"
ds = NCDataset(ncfile)

println("\n--- Dimensions ---")
for (name, dim) in ds.dim
    println("  $name: $dim")
end

println("\n--- Variables ---")
for (name, var) in ds
    println("  $name: ", size(var), "  units: ", get(var.attrib, "units", "—"))
end

# Coordinate vectors
lons   = ds["lon"][:]    # degrees_east,  2160 values
lats   = ds["lat"][:]    # degrees_north, 1080 values (N to S)
months = ds["month"][:]  # 1:12

println("\n--- Coordinate ranges ---")
println("  lon:   $(lons[1])° to $(lons[end])°  ($(length(lons)) points)")
println("  lat:   $(lats[1])° to $(lats[end])°  ($(length(lats)) points)")
println("  month: $(months[1]) to $(months[end])")

# --- Example 1: Global slice — July mean temperature ---
tmp_jul = ds["tmp"][:, :, 7]   # lon × lat at month 7
println("\n--- tmp July ---")
println("  size:  ", size(tmp_jul))
println("  range: $(round(minimum(skipmissing(tmp_jul)), digits=1)) — ",
                  "$(round(maximum(skipmissing(tmp_jul)), digits=1)) °C")

# --- Example 2: Point extraction — Alice Springs ---
target_lon, target_lat = 133.8807, -23.6980

ilon = argmin(abs.(lons .- target_lon))
ilat = argmin(abs.(lats .- target_lat))
println("\n--- Alice Springs (nearest cell: lon=$(lons[ilon])°, lat=$(lats[ilat])°) ---")

monthly_vars = ["pre", "rd0", "frs", "tmp", "dtr", "reh", "sunp", "wnd"]
for v in monthly_vars
    vals = ds[v][ilon, ilat, :]
    println("  $v: $vals")
end

elv = ds["elv"][ilon, ilat]
println("  elv: $elv m")

# Derived min/max temperature
tmp_vals = Float64.(ds["tmp"][ilon, ilat, :])
dtr_vals = Float64.(ds["dtr"][ilon, ilat, :])
tmin = tmp_vals .- dtr_vals ./ 2
tmax = tmp_vals .+ dtr_vals ./ 2
println("  tmin: $(round.(tmin, digits=1))")
println("  tmax: $(round.(tmax, digits=1))")

close(ds)

# --- Rasters.jl ---
using Rasters

# Open all variables as a RasterStack — inspect to see how dims are named
stack = RasterStack(ncfile)

# Individual variable (all months)
tmp_rast = Raster(ncfile, name = :tmp)
println("\n--- tmp Raster dims ---")
println(dims(tmp_rast))

# Global map of July temperature — index by whatever the month dim is called
# (check println above: likely Ti or Dim{:month})
tmp_jul = tmp_rast[month = At(7)]   # adjust dim name if needed
plot(tmp_jul)

# Point extraction — Alice Springs
point = stack[X(Near(target_lon)), Y(Near(target_lat))]
println("\n--- Alice Springs via Rasters.jl ---")
println(point)
